### ISSUE TITLE
Stuck in iteration #0

### ISSUE THREAD

**Issue by bibekiit** [franrruiz/p-emb#1]
It's been a day. 

The log output is 

Output directory label: t7861363-n4345806-m354433-k50-avgCtxt1-eta0.1-zF0.1-nS10
Creating directory ../model/t7861363-n4345806-m354433-k50-avgCtxt1-eta0.1-zF0.1-nS10
Initializing latent parameters...
Running inference algorithm...
 +Iteration 0...
  computing sum of alphas...
  initializing gradients...
  processing positive observations...
  processing negative observations...
  taking grad step...

Can you suggest me what parameters to take to increase the speed?

Regards,

Bibek

**Comment by franrruiz** [comment]
Some comments:

If it's stuck at "taking grad step", then there might be some numerical unstability, due to either a large stepsize, or a mispecification of the data files. Check the output file `obj_function.txt`; if you see a ridiculously large value (basically, +/-infinity), then you should reduce the stepsize (try `-eta 0.01`) or double-check the correctness of the input files.

In any case, you should definitely use minibatches of data to speed up the gradient computations. For that, use the option `-batchsize <int>`. You may consider values like {1000, 10000, 100000} (I would suggest starting with 10000 first).

Let me know how it goes.

Best,
Fran

**Comment by bibekiit** [comment]
I am getting huge gradient values even when i set eta 0.01  = 58980070279.780815
I also set batchsize to 10000. 

Can you give any hints on the correctness of input files. I have followed the format as specified in readme. Only thing is i have just divided the input file into 3 parts train, test and validation. I guess some pairs of user_id, session_id will be common in training and validation file.

**Comment by franrruiz** [comment]
That's indeed unusually large, although it doesn't seem like an overflow.

What happens with batchsize 10k (you may first try eta 0.1 for that)? Is it stuck?

**Comment by bibekiit** [comment]
At eta 0.1, batchsize 10000 I am getting grad 58980070279.780815

**Comment by franrruiz** [comment]
My question is: is it still stuck at iteration 0?

(Also, that number isn't the gradient; the value of the gradient is not written in the output)

**Comment by bibekiit** [comment]
Yes it is stuck at 0th iteration. Ok I misinterpreted it as gradient, its the value of obj function. If the obj_function value is high, will it get stuck?

**Comment by franrruiz** [comment]
In which step is it stuck? (processing negative observations, taking grad step, ...?)
If it is anywhere before "taking grad step", then it is still running and not stuck. If that's too slow for you, consider further reducing the batch size (use, e.g., 1000). The cost of each iteration is basically linear with the batch size.

**Comment by bibekiit** [comment]
Yes it is stuck at "taking gradient step". Yes now I am running with bacthsize=1000. Okay

**Comment by franrruiz** [comment]
That's weird. There might be something wrong with reading the input files; please read carefully the readme and make sure that the input files are in the right format, and that you indeed have 4.3M users, 350K items, and 7.8M baskets.

Besides that, I can only think of another two sanity checks. One of them is to use `-bernoulli`. In my experience the Bern likelihood is numerically more stable than the Poisson. The other one is to set `-eta 0`. This basically means "don't update the parameters", i.e., no optimization is performed. Does it also get stuck with either of these options?

**Comment by franrruiz** [comment]
By the way, do you have a test/validation set? If these files are way too large, then it might be taking a lot of time to obtain test/validation performance. Can you use `-noVal -noTest` and see what happens?

**Comment by franrruiz** [comment]
Actually my best bet now is that the issue is due to a test/validation set that's too large.

**Comment by bibekiit** [comment]
Lines of train.tsv=32369055
 +Lines of test.tsv=9248301
 +Lines of validation.tsv=13872452

**Comment by bibekiit** [comment]
It is running with -noVal -noTest. What should be the size of test file for it to run easily.

**Comment by bibekiit** [comment]
I got this error 
*** stack smashing detected ***: ./pemb terminated

**Comment by bibekiit** [comment]
These are my objective values with bernoulli eta = 0.1 batchsize =10000


0       -5924460389.977258
1       -5926737941.822274
2       -5848591943.151537
3       -5226717562.509828
4       -3796153547.260298
5       -2260163141.063970
6       -1340277342.494691
7       -893653655.884120
8       -666230112.226719
9       -533745268.050134
10      -450191768.171854
11      -401830368.478109
12      -362660229.921621
13      -331697180.727831
14      -310429789.176455
15      -292780800.518785
16      -276609010.418343
17      -265369290.724612
18      -258016611.063155
19      -249284693.269456
20      -242096679.006752
21      -235074031.532052
22      -231843136.589586
23      -224611609.747266
24      -218628455.459818
25      -217120801.141664
26      -212471313.899160
27      -209701228.916595
28      -207536686.844760
29      -205192157.796361
30      -202718780.391966
31      -199922752.910949
32      -196526960.985941
33      -194285238.006392
34      -193998895.630407
35      -193068059.917832
36      -191107948.038991
37      -187004028.353443
38      -189233326.812773
39      -188097448.160730
40      -185006791.705649
41      -184352775.947714
42      -183969177.254803
43      -183302371.146633
44      -181689356.914051
45      -180690944.400670
46      -179010664.928190
47      -178211923.310198
48      -177894460.911454
49      -180764431.563507
50      -174436995.215923
51      -175161039.542801
52      -174481884.199879
53      -174414000.576054
54      -173615840.868589
55      -173873765.245587
56      -172148144.910158
57      -173779617.106307
58      -171687936.671683
59      -172413862.522277
60      -169042885.451944
61      -170598131.947855
62      -170463829.817566
63      -169072747.301932
64      -168260859.836015
65      -167142441.274324
66      -168201785.069231
67      -169373985.586495
68      -168597984.254978
69      -168676913.245601


Such high values. Should I increase batchsize?

**Comment by franrruiz** [comment]
re size of test files: not sure, you may try different sizes.

re stack smashing: it looks like that particular experiment had a value of `-eta` too large. I can see that you finally solved the problem.

re objective values: No, they aren't excessively high values; they are ok. And the objective function is clearly increasing, so the optimization is going well. I would say these values of eta and batchsize are kind of adequate, but increasing the batchsize never hurts.
